Update config doc with Java client (#11143)
aspnet/Docs: aspnetcore/signalr/configuration.md

## Configure client options

> Client options can be configured on the `HubConnectionBuilder` type (available in the .NET and Javascipt clients. It is also available on the Java client but the `HttpHubConnectionBuilder` subclass is what contains the builder configuation options.), as well as on the `HubConnection` itself.

Reviewer: ```suggestion
Client options can be configured on the `HubConnectionBuilder` type (available in the .NET and JavaScript clients). It's also available in the Java client, but the `HttpHubConnectionBuilder` subclass is what contains the builder configuration options, as well as on the `HubConnection` itself.
```

Reviewer 2: :+1: I also like removing the parens entirely:

```suggestion
Client options can be configured on the `HubConnectionBuilder` type available in the .NET and JavaScript clients. It's also available in the Java client, but the `HttpHubConnectionBuilder` subclass is what contains the builder configuration options, as well as on the `HubConnection` itself.
```

    .build();
```

> See [Logging in SignalR](xref:signalr/java-client#Add-logging) for how to enable logging in the Java client

Reviewer: ```suggestion
For details on how to enable logging in the Java client, see [Logging in SignalR](xref:signalr/java-client#Add-logging).
```

| `Debug` | Diagnostic messages useful for debugging. |
| `Trace` | Very detailed diagnostic messages designed for diagnosing specific issues. |

> For details on how to enable logging in the Java client, see [Logging in SignalR](xref:signalr/java-client#Add-logging).

Reviewer 3: Probably not in scope for this PR, but this sentence is kind of hidden, we might want to rework the logging stuff at some point

Author: I'm just gonna copy over the logging instructions from the other page. Better to just have the info here to match the other sections.

    .build();
```

> In the Java client the transport is selected with the `withTransport` method on the HttpHubConnectionBuilder. The Java client defaults to using the WebSockets transport.

Reviewer 3: Need to moniker this for 3.0 only

> In the Java client the transport is selected with the `withTransport` method on the HttpHubConnectionBuilder. The Java client defaults to using the WebSockets transport.

Reviewer: ```suggestion
In the Java client, the transport is selected with the `withTransport` method on the `HttpHubConnectionBuilder`. The Java client defaults to using the WebSockets transport.
```

```java

>         HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/myhub")

Reviewer: dedent this code snippet

    .build();
```

::: moniker range=">= aspnetcore-2.2"

In this version of the Java client websockets is the only available transport.

::: moniker-end

::: moniker range="= aspnetcore-3.0"

In the Java client, the transport is selected with the `withTransport` method on the `HttpHubConnectionBuilder`. The Java client defaults to using the WebSockets transport.

```java
HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/myhub")
    .withTransport(TransportEnum.WEBSOCKETS)
    .build();
```

::: moniker-end

> [!NOTE]

> > The SignalR Java client doesn't support transport fallback yet.

Reviewer 3: Move this into the 3.0 range

    .build();
```

> In the SignalR Java client, you can configure a bearer token to use for authentication by providing an "access token factory" to the [HttpHubConnectionBuilder](/java/api/com.microsoft.signalr._http_hub_connection_builder?view=aspnet-signalr-java). Use [withAccessTokenFactory](/java/api/com.microsoft.signalr._http_hub_connection_builder.withaccesstokenprovider?view=aspnet-signalr-java#com_microsoft_signalr__http_hub_connection_builder_withAccessTokenProvider_Single_String__) to provide an [RxJava](https://github.com/ReactiveX/RxJava) [Single<String>](http://reactivex.io/documentation/single.html). With a call to [Single.defer](http://reactivex.io/RxJava/javadoc/io/reactivex/Single.html#defer-java.util.concurrent.Callable-), you can write logic to produce access tokens for your client.

Reviewer 2: Why put quotes around "access token factory"?

Author: I copy pasted this from somewhere else. Quotes not needed

```java
HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/myhub")
    .withAccessTokenProvider(Single.defer(() -> {
        // Your logic here.
        return Single.just("An Access Token");
    })).build();
```

### Configure timeout and keep-alive options

Additional options for configuring timeout and keep-alive behavior are available on the `HubConnection` object itself:

> | .NET Option | JavaScript Option | Java Option | Default Value | Description |

Reviewer 3: We'll probably want to refactor this table now that it's getting bigger.

Reviewer: I agree with Brennan. What do you think about using 3 tabs: .NET, JavaScript, and Java?

Author: > 3 tabs: .NET, JavaScript, and Java?

Just want to clarify,  under Configuration in the TOC having separate sections for each client? Or something else?

Reviewer: I'm imagining actual tabs (see https://docs.microsoft.com/aspnet/core/tutorials/first-web-api?view=aspnetcore-2.2&tabs=visual-studio#create-a-web-project for an example). The syntax for the tabs would be:

```markdown
[.NET](#tab/dotnet)

.NET-specific table here

[JavaScript](#tab/javascript)

JavaScript-specific table here

[Java](#tab/java)

Java-specific table here

---
```

Author: Oh, awesome. Thanks for the clarification. 👍

Reviewer 4: ver ver cool

Reviewer 3: Remove this section now

Reviewer: I agree with Brennan's suggestion ☝️ . Also remove the 2 sentences below this table.

| ----------- | ----------------- | ----------- | ------------- | ----------- |
| `ServerTimeout` | `serverTimeoutInMilliseconds` | `getServerTimeout` `setServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |

> | `HandshakeTimeout` | Not configurable |  `withHandshakeResponseTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript). This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

Reviewer 2: > If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript)

Make sure to mention `onClosed` for Java.

> In the .NET Client, timeout values are specified as `TimeSpan` values. In the JavaScript client, timeout values are specified as a number indicating the duration in milliseconds.

Reviewer 2: What type is used f Java client timeouts?

Author: `long`s representing milliseconds. I'll add it

| ----------- | ----------------- | ----------- | ------------- | ----------- |
| `AccessTokenProvider` | `accessTokenFactory` | `withAccessTokenProvider` | `null` | A function returning a string that is provided as a Bearer authentication token in HTTP requests. |
| `SkipNegotiation` | `skipNegotiation` | `shoudlSkipNegotiate` | `false` | Set this to `true` to skip the negotiation step. **Only supported when the WebSockets transport is the only enabled transport**. This setting can't be enabled when using the Azure SignalR Service. |
| `ClientCertificates` | Not configurable * | Empty | A collection of TLS certificates to send to authenticate requests. |
| `Cookies` | Not configurable * | Not configurable | Empty | A collection of HTTP cookies to send with every HTTP request. |
| `Credentials` | Not configurable * | Not configurable | Empty | Credentials to send with every HTTP request. |
| `CloseTimeout` | Not configurable * | Not configurable | 5 seconds | WebSockets only. The maximum amount of time the client waits after closing for the server to acknowledge the close request. If the server doesn't acknowledge the close within this time, the client disconnects. |

> | `Headers` | Not configurable * | `withHeader``withHeaders` | Empty | A dictionary of additional HTTP headers to send with every HTTP request. |

Reviewer 2: I wonder if `Map` would be more familiar than dictionary to Java devs.

| `HandshakeTimeout` | Not configurable |  `withHandshakeResponseTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript). This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

In the .NET Client, timeout values are specified as `TimeSpan` values. In the JavaScript client, timeout values are specified as a number indicating the duration in milliseconds.

> [.NET](#tab/dotnet)

Reviewer: ```suggestion
# [.NET](#tab/dotnet)

```

> | .NET Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

| ----------- | ------------- | ----------- |
| `ServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `Closed` event (`onclose` in JavaScript). This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |
| `HandshakeTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript). This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

In the .NET Client, timeout values are specified as `TimeSpan` values.

> [Javascript](#tab/javascript)

Reviewer: ```suggestion
# [JavaScript](#tab/javascript)

```

> | JavaScript Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

| ----------------- | ------------- | ----------- |
 `serverTimeoutInMilliseconds` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `onclose` event. This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |

>  [Java](#tab/java)

Reviewer: ```suggestion
# [Java](#tab/java)

```

> | Java Option | Default Value | Description |

Reviewer: ```suggestion
| Option | Default value | Description |
```

| ----------- | ------------- | ----------- |
|`getServerTimeout` `setServerTimeout` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `onClose` event. This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |
| `withHandshakeResponseTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `onClose` event. This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

> 

Reviewer: Terminate the tab set with 3 dashes.

```suggestion

---

```

| `HandshakeTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `Closed` event (`onclose` in JavaScript). This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

In the .NET Client, timeout values are specified as `TimeSpan` values.

# [JavaScript](#tab/javascript)

| Option | Default value | Description |
| ----------------- | ------------- | ----------- |

>  `serverTimeoutInMilliseconds` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `onclose` event. This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |

Reviewer: ```suggestion
| `serverTimeoutInMilliseconds` | 30 seconds (30,000 milliseconds) | Timeout for server activity. If the server hasn't sent a message in this interval, the client considers the server disconnected and triggers the `onclose` event. This value must be large enough for a ping message to be sent from the server **and** received by the client within the timeout interval. The recommended value is a number at least double the server's `KeepAliveInterval` value, to allow time for pings to arrive. |

In the JavaScript client, timeout values are specified as a number indicating the duration in milliseconds.
```

| `withHandshakeResponseTimeout` | 15 seconds | Timeout for initial server handshake. If the server doesn't send a handshake response in this interval, the client cancels the handshake and triggers the `onClose` event. This is an advanced setting that should only be modified if handshake timeout errors are occurring due to severe network latency. For more detail on the Handshake process, see the [SignalR Hub Protocol Specification](https://github.com/aspnet/SignalR/blob/master/specs/HubProtocol.md). |

---

### Configure additional options

Additional options can be configured in the `WithUrl` (`withUrl` in JavaScript) method on `HubConnectionBuilder` or on the various configuration APIs on the `HttpHubConnectionBuilder` in the Java client:

> | .NET Option | JavaScript Option | Java Option | Default Value | Description |

Reviewer 3: Tabs

| ----------- | ----------------- | ----------- | ------------- | ----------- |
| `AccessTokenProvider` | `accessTokenFactory` | `withAccessTokenProvider` | `null` | A function returning a string that is provided as a Bearer authentication token in HTTP requests. |
| `SkipNegotiation` | `skipNegotiation` | `shouldSkipNegotiate` | `false` | Set this to `true` to skip the negotiation step. **Only supported when the WebSockets transport is the only enabled transport**. This setting can't be enabled when using the Azure SignalR Service. |
| `ClientCertificates` | Not configurable * | Empty | A collection of TLS certificates to send to authenticate requests. |
| `Cookies` | Not configurable * | Not configurable | Empty | A collection of HTTP cookies to send with every HTTP request. |
| `Credentials` | Not configurable * | Not configurable | Empty | Credentials to send with every HTTP request. |
| `CloseTimeout` | Not configurable * | Not configurable | 5 seconds | WebSockets only. The maximum amount of time the client waits after closing for the server to acknowledge the close request. If the server doesn't acknowledge the close within this time, the client disconnects. |

> | `Headers` | Not configurable * | `withHeader``withHeaders` | Empty | A Map of additional HTTP headers to send with every HTTP request. |

Reviewer 3: `withHeader``withHeaders` needs a space

    .build();
```

In the Java client, these options can be configured with the methods on the `HttpHubConnectionBuilder` returned from the `HubConnectionBuilder.create("HUB URL")`

```java

>     HubConnection hubConnection = HubConnectionBuilder.create("https://example.com/myhub")

Reviewer: dedent this code snippet

            .withHeader("Foo", "Bar")

>             .shoudldSkipNegotiate(true)

Reviewer 3: shouldSkipNegotiate
